docs: persist and connect state with url (#1804)
pmndrs/zustand: docs/guides/connect-to-state-with-url-hash.md

)
```

### CodeSandbox Demo

> https://codesandbox.io/s/zustand-state-with-url-hash-demo-f29b88?file=/src/store/index.ts

Reviewer: ```suggestion
<https://codesandbox.io/s/zustand-state-with-url-hash-demo-f29b88?file=/src/store/index.ts>
```

## Persist and Connect State with URL Parameters (Example: URL Query Parameters)

> This example is for when you want to conditionally connect the state to the URL (here we use URL query parameters) when visiting the website from a populated URL, and always want it synced and consistent with something like localstorage.

Reviewer: You can split the long lines according to <https://sembr.org/>

```
This example is for when you want to conditionally connect the state to the URL
(here we use URL query parameters) when visiting the website from a populated URL,
and always want it synced and consistent with something like `localstorage`.
```

However, I'd rephrase this intro:

```suggestion
There are times when you want to conditionally connect the state to the URL.
This example depicts usage of the URL query parameters
while keeping it synced with another persistence implementation, like `localstorage`.
```

If you want the URL params to always populate, the conditional check on `window.location.search.slice(1)` can be removed.

The implementation below will update the URL in place, without refresh, as the relevant states change.

```ts
import { create } from 'zustand'
import { persist, StateStorage, createJSONStorage } from 'zustand/middleware'

const persistentStorage: StateStorage = {
  getItem: (key): string => {
    // Check URL first
    if (window.location.search.slice(1)) {

>       const searchParams = new URLSearchParams(window.location.search.slice(1))

Reviewer: ```js
      const searchParams = new URLSearchParams(window.location.search.slice(1))
      // consider extracting to a function ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

      // so that it could shortened into:
      new URLSearchParams(windowSearchParams())
```

      const storedValue = searchParams.get(key)

>       return JSON.parse(storedValue) // **

Reviewer: ```js
      return JSON.parse(storedValue) // **
      // this comment is unclear -------^^
```

    } else {
      // Otherwise, we should load from localstorage or alternative storage
      return JSON.parse(localStorage.getItem(key))
    }
  },
  setItem: (key, newValue): void => {
    // Check if query params exist at all, can remove check if always want to set URL

>     if (window.location.search.slice(1)) {

Reviewer: ```js
    if (window.location.search.slice(1)) {
    //  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ --- replace other ocurrences with extracted method
```

      const searchParams = new URLSearchParams(window.location.search.slice(1))

>       searchParams.set(key, JSON.stringify(newValue)) // **

Reviewer: ```js
      searchParams.set(key, JSON.stringify(newValue)) // **
      // this comment is also unclear -------------------^^
```

      window.history.replaceState(null, null, `?${searchParams.toString()}`)
    }

    localStorage.setItem(key, JSON.stringify(newValue))
  },
  removeItem: (key): void => {
    const searchParams = new URLSearchParams(window.location.search.slice(1))
    searchParams.delete(key)
    window.location.search = searchParams.toString()
  },
}

let localAndUrlStore = (set) => ({
  searchGameDifficulty: {
    beginner: false,
    intermediate: true,
    expert: false,
  },
  setSearchGameDifficulty: (difficulty, selection) =>
    set((state) => ({
      searchGameDifficulty: {
        ...state.searchGameDifficulty,
        [difficulty]: selection,
      },
    })),

  searchRatings: {
    1: false,
    2: false,
    3: false,
    4: true,
    5: true,
  },
  setSearchRatings: (rating, selection) =>
    set((state) => ({
      searchRatings: { ...state.searchRatings, [rating]: selection },

>     })),

Reviewer: Could you prepare a simpler example? It took me a fair amount of time to understand the domain of the problem. Something obvious would be more suitable, because the example needs to depict the state in URL, not the game search problem.

})

let storageOptions = {
  name: 'gameSearchPreferences',
  storage: persistentStorage,
}

const useLocalAndUrlStore = create(

>   persist(devtools(localAndUrlStore), storageOptions)

Reviewer: ```js
  persist(devtools(localAndUrlStore), storageOptions)
  //      ^^^^^^^^^----------------^ --- you can drop the devtools, as they are not required for this example
```

)

export default localAndUrlStore
```

When generating the URL from a component, you can call buildShareableUrl:

```ts
const buildURLSuffix = (params, version = 0) => {
  const searchParams = new URLSearchParams()

  const zustandStoreParams = {
    state: {
      searchGameDifficulty: params.searchGameDifficulty,
      searchRatings: params.searchRatings,
    },
    version: version, // version is here because that is included with how Zustand sets the state
  }

  // The URL param key should match the name of the store, as specified as in storageOptions above
  searchParams.set('gameSearchPreferences', JSON.stringify(zustandStoreParams)) // **
  return searchParams.toString()
}

export const buildShareableUrl = (params, version) => {
  return `${window.location.origin}?${buildURLSuffix(params, version)}`
}
```

The generated URL would look like (here without any encoding, for readability):

> https://localhost/search?gameSearchPreferences={"state":{"searchGameDifficulty":{"beginner":false,"intermediate":true,"expert":false},"searchRatings":{"1":false,"2":false,"3":false,"4":true,"5":true}},"version":0}}

Reviewer: ```suggestion
\```
https://localhost/search?gameSearchPreferences={"state":{"searchGameDifficulty":{"beginner":false,"intermediate":true,"expert":false},"searchRatings":{"1":false,"2":false,"3":false,"4":true,"5":true}},"version":0}}
\```
```

Add code fences for better link readability, but without the `\` :)
